pepper-apis/ddp-common/src/main/java/org/broadinstitute/ddp/service/FileDownloadService.java

```diff
@@ -0,0 +1,86 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+public class FileDownloadService {
+    private final ServiceAccountSigner signer;
+    private final GoogleBucketClient storageClient;
+    private final int maxSignedUrlMins;
+    private final String defaultDownloadBucket;
+
+    //GCP Service Account for file download
+    public static final String FILE_DOWNLOAD_CREDENTIALS = "fileDownload.signerServiceAccount";
+    public static final String FILE_DOWNLOAD_PROJECT_ID = "fileDownload.signerServiceAccount.project_id";
+    public static final String MAX_SIGN_URL_MIN = "fileDownload.maxSignedUrlMins";
+    public static final String DEFAULT_DOWNLOAD_BUCKET = "fileDownload.defaultDownloadBucket";
+
+    public static FileDownloadService fromConfig(Config cfg) {
+
+        String signerJson = ConfigUtil.toJson(cfg.getConfig(FILE_DOWNLOAD_CREDENTIALS));
+        InputStream signerStream = new ByteArrayInputStream(signerJson.getBytes(StandardCharsets.UTF_8));
+        ServiceAccountCredentials signerCredentials;
+        try {
+            signerCredentials = ServiceAccountCredentials.fromStream(signerStream);
+        } catch (IOException e) {
+            throw new DDPException("Could not get signer credentials", e);
+        }
+
+        GoogleCredentials bucketCredentials;
+        boolean ensureDefault = cfg.getBoolean(ConfigFile.REQUIRE_DEFAULT_GCP_CREDENTIALS);
+        bucketCredentials = GoogleCredentialUtil.initCredentials(ensureDefault);
+        if (bucketCredentials == null) {
+            log.error("Could not get bucket credentials, defaulting to signer credentials");
+            bucketCredentials = signerCredentials;
+        }
+
+        String projectId = cfg.getString(FILE_DOWNLOAD_PROJECT_ID);
+
+        return new FileDownloadService(
+                signerCredentials,
+                new GoogleBucketClient(projectId, bucketCredentials),
+                cfg.getInt(MAX_SIGN_URL_MIN), cfg.getString(DEFAULT_DOWNLOAD_BUCKET));
+    }
+
+    public FileDownloadService(ServiceAccountSigner signer, GoogleBucketClient storageClient,
+                               int maxSignedUrlMins, String defaultDownloadBucket) {
+        this.signer = signer;
+        this.storageClient = storageClient;
+        this.maxSignedUrlMins = maxSignedUrlMins;
+        this.defaultDownloadBucket = defaultDownloadBucket;
+    }
+
+    /**
+     * Generating a signed URL for file download
+     *
+     * @param fileName   the name of the file
+     * @param bucketName the name of the bucket
+     * @return URL
+     */
+    public URL getSignedURL(String fileName, String bucketName) {
+        HttpMethod method = HttpMethod.GET;
+        return storageClient.generateSignedUrl(
+                signer, StringUtils.isEmpty(bucketName) ? defaultDownloadBucket : bucketName, fileName,
```

Comment:
If it's empty, I wouldn't go to a default bucket configuration.  I'd throw an exception at this point because we can't afford to connect a user to the wrong file.

Reply:
As of now assuming top level bucket will be same for results/download files and not expecting to pass for each request. Its more of. for any future use.
Assuming DSM will pass the full patch that is specific to each participant results_file with each on-demand trigger. we can discuss more if needed

Reply:
DSM needs to provide the bucket and blob.  full stop.

That being said, there's nothing to guarantee that the blob will still exist when you go to create the signed URL, because DSM also has a feature to delete one of these documents.  Not sure if it makes sense to check to see if the bucket/blob still exist before returning the signed URL.  Happy to discuss this further.

Reply:
From my understanding for launch we wont handle any deletions. DSS side deletions will be data change requests to hide that appropriate activity instance.
If the file doesn't exist it will eventually fail when the user tries to download.

Reply:
PEPPER-67 requires DSM handle deletion at day 1.  This doesn't mean the activity needs to be cleaned up.  We just need to realize the download link may point to an invalid file.  You have the opportunity to check and see if the file exists at this point and could return something so the client can decide to do something different.  Maybe  a getBlob(bucket, blobPath).exists() and return a 404 if it's been deleted?

Reply:
Updated code already for File check